pkg/gadgets/internal/test/helpers.go

```diff
@@ -0,0 +1,161 @@
+//go:build linux
+// +build linux
+
+// Copyright 2022 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package test
+
+import (
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/cilium/ebpf"
+	"github.com/google/go-cmp/cmp"
+	"golang.org/x/sys/unix"
+)
+
+// CreateMntNsFilterMap creates and fills an eBPF map that can be used
+// to filter by mount namespace in the different tracers.
+func CreateMntNsFilterMap(t *testing.T, mountNsIDs ...uint64) *ebpf.Map {
+	t.Helper()
+
+	const one = uint32(1)
+
+	mntnsSpec := &ebpf.MapSpec{
+		Name:       "filter_map_" + t.Name(),
+		Type:       ebpf.Hash,
+		KeySize:    8,
+		ValueSize:  4,
+		MaxEntries: 1024,
+	}
+	m, err := ebpf.NewMap(mntnsSpec)
+	if err != nil {
+		t.Fatalf("Failed to create eBPF map: %s", err)
+	}
+
+	for _, mountnsid := range mountNsIDs {
+		if err := m.Put(mountnsid, one); err != nil {
+			m.Close()
+			t.Fatalf("Failed to update eBPF map: %s", err)
+		}
+	}
+
+	return m
+}
+
+// RequireRoot skips the test if the not running as root
+func RequireRoot(t *testing.T) {
+	t.Helper()
+	if unix.Getuid() != 0 {
+		t.Skip("Test requires root")
+	}
+}
+
+type Tracer interface {
+	Stop()
+}
+
+type ValidateEventType[Event any, Ret any] func(t *testing.T, info *Info, ret *Ret, events []Event)
+
+type Test[Event any, Ret any] struct {
+	GetTracer     func(*Info, func(event Event)) Tracer
+	RunnerConfig  *Config
+	GenerateEvent func() (Ret, error)
+	ValidateEvent ValidateEventType[Event, Ret]
+}
+
+// Run runs a series of test cases using a dedicated runner.
+func Run[Event any, Ret any](t *testing.T, tests map[string]Test[Event, Ret]) {
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
```

Comment:
I really like the idea of runner with a separate goroutine, but it seems to me this code takes away a lot of flexibility from the tests and does not allow proper, fine-grained testing. If we inline sample test config with this code and do some optimization, we could get to something like below. No generics and I think dramatically simpler code, even if between the tests there will be some duplication. But duplication in tests is actually more acceptable, as it makes tests easier to follow. I"m curious what are your thoughts on such approach.
```go
			config := &tracer.Config{
				MountnsMap: utilstest.CreateMntNsFilterMap(t, runner.Info.MountNsID),
			}

			tracer, events := newTracer(t, config) // This would be test helper function defined per tracer with tracer specific types.

      pid := 0

			utilstest.RunWithRunner(t, &utilstest.Config{UID: 1435}, func(){
				cmd := exec.Command("/bin/cat", "/dev/null")
				if err := cmd.Run(); err != nil {
					t.Fatalf("Error running command: %v", err)
				}

        // We can export any information from the runner as needed.
				pid = cmd.Process.Pid
			})

			// Give some time for the tracer to capture the events
			time.Sleep(100 * time.Millisecond)

			if len(events) != 0 {
				t.Fatalf("bla")
			}
```

Reply:
It's a good question. IMO it's better to have a table based testing and "hide" some of the logic in the helpers.go. It reduces duplication and makes it faster to implement tests for other gadgets.

I tried to port some of the tests and it's the result: (after making runner public...)

```golang
func TestTracer2(t *testing.T) {
	events := []types.Event{}
	eventCallback := func(event types.Event) {
		events = append(events, event)
	}

	runner, err := utilstest.NewRunner(&utilstest.RunnerConfig{})
	if err != nil {
		t.Fatalf("%q: failed to create runner", err)
	}
	t.Cleanup(runner.Close)

	config := &tracer.Config{
		MountnsMap: utilstest.CreateMntNsFilterMap(t, runner.Info.MountNsID),
	}

	tracer := createTracer(t, config, eventCallback)
	t.Cleanup(tracer.Stop)

	var catPid int

	runner.Run(func() error {
		cmd := exec.Command("/bin/cat", "/dev/null")
		if err := cmd.Run(); err != nil {
			return fmt.Errorf("running command: %w", err)
		}

		catPid = cmd.Process.Pid

		return nil
	})

	// Give some time for the tracer to capture the events
	time.Sleep(100 * time.Millisecond)

	if len(events) != 1 {
		t.Fatalf("more than one event found")
	}

	expectedEvent := &types.Event{
		Event: eventtypes.Event{
			Type: eventtypes.NORMAL,
		},
		Pid:       uint32(catPid),
		Ppid:      uint32(runner.Info.Pid),
		UID:       uint32(runner.Info.UID),
		MountNsID: runner.Info.MountNsID,
		Retval:    0,
		Comm:      "cat",
		Args:      []string{"/bin/cat", "/dev/null"},
	}

	if !reflect.DeepEqual(expectedEvent, &events[0]) {
		t.Fatalf("event doesn't match:\n%s",
			cmp.Diff(expectedEvent, &events[0]))
	}
}
```

IMO we can continue with this PR as it's. My main goal was to implement the runner to emulate events happening inside containers, and it's not so difficult to change from a table base approach to the suggested here in the future if we really need to. (Also it could work having a mix, small tests could use the table base approach and more elaborated ones could require to create the handle the runner).

Reply:
I think with:
```go
	events := []types.Event{}
	eventCallback := func(event types.Event) {
		events = append(events, event)
	}

	runner, err := utilstest.NewRunner(&utilstest.RunnerConfig{})
	if err != nil {
		t.Fatalf("%q: failed to create runner", err)
	}
	t.Cleanup(runner.Close)
```
simplified to:
```go
	events, callback := createCallback()

	runner, err := utilstest.NewRunner(t, &utilstest.RunnerConfig{})
```

And with part:
```go
	expectedEvent := &types.Event{
		Event: eventtypes.Event{
			Type: eventtypes.NORMAL,
		},
		Pid:       uint32(catPid),
		Ppid:      uint32(runner.Info.Pid),
		UID:       uint32(runner.Info.UID),
		MountNsID: runner.Info.MountNsID,
		Retval:    0,
		Comm:      "cat",
		Args:      []string{"/bin/cat", "/dev/null"},
	}

	if !reflect.DeepEqual(expectedEvent, &events[0]) {
		t.Fatalf("event doesn't match:\n%s",
			cmp.Diff(expectedEvent, &events[0]))
	}
```

reduced to assertion on expected fields (or one test can do full comparison, but definitely not all of them IMO), I think this actually gives really nice result.

Reply:
I think the approach I propose is more loosely coupled and you are less likely to run into overly customizable test API once it turns out that other tracers require some extra setup or something else. With this approach, you make easy thing easy and hard thing possible I think.

Reply:
Also, I'd argue that it's better to duplicate things at first and see where exactly the patterns emerge, then try to make a "framework" out of it, rather than starting with something structurally tight from the beginning. I think runner is a great example of that, it's pluggable instead of enforcing the convention.

Reply:
I gave it a try. What do you think of the result?

Reply:
Looks good now, thanks!